File: app/src/main/java/com/example/android/xo/GameBoard.java
package com.example.android.xo;

import android.content.res.Resources;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.example.android.xo.Constants.Constants;

import java.util.concurrent.TimeUnit;

import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import ir.tapsell.plus.AdRequestCallback;
import ir.tapsell.plus.AdShowListener;
import ir.tapsell.plus.TapsellPlus;
import ir.tapsell.plus.TapsellPlusBannerType;
import ir.tapsell.plus.model.TapsellPlusAdModel;
import ir.tapsell.plus.model.TapsellPlusErrorModel;

public class GameBoard extends AppCompatActivity {

    public static final int x_player = 12;
    public static final int o_player = 15;
    public static final int empty = 20;
    public static final int No_Winner = -1;
    int winner = No_Winner;
    int[] status = {empty, empty, empty, empty, empty, empty, empty, empty, empty};
    int active_player;
    int[][] winner_state = {{0, 1, 2}, {3, 4, 5}, {6, 7, 8}, {0, 3, 6}, {1, 4, 7}, {2, 5, 8}, {0, 4, 8}, {2, 4, 6}};

    int x_points = 0;
    int o_points = 0;

    ImageView[] columns;

    private Long backPressedTime = 0L;
    TextView current;
    TextView x_point;
    TextView o_point;
    Resources resources;
    FrameLayout adContainer;

    private String standardBannerResponseId = "";
    private final CompositeDisposable disposable = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.board);
        resources = getResources();
        adContainer = findViewById(R.id.standard_ad_container);

        columns = new ImageView[]{findViewById(R.id.c0), findViewById(R.id.c1), findViewById(R.id.c2),
                findViewById(R.id.c3), findViewById(R.id.c4), findViewById(R.id.c5),
                findViewById(R.id.c6), findViewById(R.id.c7), findViewById(R.id.c8)};

        current = findViewById(R.id.current_player);
        x_point = findViewById(R.id.x_point);
        o_point = findViewById(R.id.o_point);

        startGame();

        x_point.setText(String.format(resources.getString(R.string.x_point), 0));
        o_point.setText(String.format(resources.getString(R.string.o_point), 0));
        requestTapsellStandard();
    }

    public void click(View view) {
        ImageView image = (ImageView) view;
        int num = Integer.parseInt((String) view.getTag());
        if (status[num] != empty || winner != No_Winner)
            return;
        image.setAlpha(0f);
        if (active_player == o_player) {
            image.setImageResource(R.drawable.o);
            active_player = x_player;
            status[num] = o_player;
        } else if (active_player == x_player) {
            image.setImageResource(R.drawable.x);
            active_player = o_player;
            status[num] = x_player;
        }
        image.animate().alpha(1f).setDuration(1000);
        winner = checkWinner();
        if (winner != No_Winner) {
            Toast.makeText(this, "winner : " + ((winner == x_player) ? "x player" : "o player"), Toast.LENGTH_SHORT).show();
        }
        if (filled() || winner != No_Winner) {
            plus_point(winner);
            disposable.add(Completable
                    .timer(1, TimeUnit.SECONDS, AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe(this::showDialogGame));
        }
        changeCurrentText();
    }

    private void showDialogGame() {
        AlertDialog.Builder builder1 = new AlertDialog.Builder(this);
        builder1.setMessage("Another Game?");
        builder1.setCancelable(false);

        builder1.setPositiveButton(
                "Continue",
                (dialog, id) -> finishGame());

        builder1.setNegativeButton(
                "Exit",
                (dialog, id) -> {
                    finish();
                    dialog.cancel();
                });

        AlertDialog alert11 = builder1.create();
        alert11.show();
    }

    private void clearMap() {
        for (int j = 0; j < 9; j++) {
            status[j] = empty;
            columns[j].setAlpha(0f);
        }
        winner = No_Winner;
    }

    private void finishGame() {
        clearMap();
        startGame();
    }

    @Override
    protected void onDestroy() {
        destroyAd();
        super.onDestroy();
    }

    private void startGame() {
        if (Math.random() < 0.5)
            active_player = x_player;
        else
            active_player = o_player;
        Toast.makeText(this, "first player is : " + (active_player == x_player ? "X" : "O"), Toast.LENGTH_LONG).show();

        changeCurrentText();
    }

    private void changeCurrentText() {
        if (active_player == x_player)
            current.setText(R.string.current_x);
        else
            current.setText(R.string.current_o);
    }

    private int checkWinner() {
        for (int[] post : winner_state) {
            if (status[post[0]] == status[post[1]] && status[post[1]] == status[post[2]] && status[post[0]] != empty) {
                return status[post[0]];
            }
        }
        return No_Winner;
    }

    private boolean filled() {
        for (int j : status) {
            if (j == empty)
                return false;
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if (backPressedTime + 3000 > System.currentTimeMillis()) {
            super.onBackPressed();
            finish();
        } else {
            Toast.makeText(this, "Press back again to leave the app.", Toast.LENGTH_LONG).show();
        }
        backPressedTime = System.currentTimeMillis();
    }

    private void plus_point(int player) {
        if (player == x_player) {
            x_points += 1;
            x_point.setText(String.format(resources.getString(R.string.x_point), x_points));
        } else if (player == o_player) {
            o_points += 1;
            o_point.setText(String.format(resources.getString(R.string.o_point), o_points));
        }
    }

    private void requestTapsellStandard() {
        TapsellPlus.requestStandardBannerAd(
                this, Constants.STANDARD_ZONE_ID,
                TapsellPlusBannerType.BANNER_320x50,
                new AdRequestCallback() {
                    @Override
                    public void response(TapsellPlusAdModel tapsellPlusAdModel) {
                        super.response(tapsellPlusAdModel);

                        standardBannerResponseId = tapsellPlusAdModel.getResponseId();
                        Log.i("Tapsell", "standard banner request created");
                    }

                    @Override
                    public void error(@NonNull String message) {
                        adContainer.setVisibility(View.GONE);
                        Log.e("Tapsell", "standard banner request got error");
                    }
                });

        disposable.add(Completable
                .timer(6, TimeUnit.SECONDS, AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(this::showStandardAd));
    }

    private void showStandardAd() {
        TapsellPlus.showStandardBannerAd(this, standardBannerResponseId,
                adContainer,
                new AdShowListener() {
                    @Override
                    public void onOpened(TapsellPlusAdModel tapsellPlusAdModel) {
                        Log.i("Tapsell", "standard banner opened");
                        adContainer.setVisibility(View.VISIBLE);
                        super.onOpened(tapsellPlusAdModel);
                    }

                    @Override
                    public void onError(TapsellPlusErrorModel tapsellPlusErrorModel) {
                        Log.i("Tapsell", "standard banner error on show");
                        adContainer.setVisibility(View.GONE);
                        super.onError(tapsellPlusErrorModel);
                    }
                });
    }

    private void destroyAd() {
        TapsellPlus.destroyStandardBanner(this, standardBannerResponseId, findViewById(R.id.standard_ad_container));
    }
}

File: app/src/main/java/com/example/android/xo/GameBoard.kt
package com.example.android.xo

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import com.example.android.xo.audio.HapticManager
import com.example.android.xo.audio.SoundManager
import com.example.android.xo.data.GamePreferences
import com.example.android.xo.engine.CellState
import com.example.android.xo.engine.GameMode
import com.example.android.xo.engine.GameResult
import com.example.android.xo.engine.Player
import com.example.android.xo.engine.TicTacToeAi
import com.example.android.xo.engine.TicTacToeGame
import com.example.android.xo.ui.WinningLineView
import com.example.android.xo.util.WindowInsetsUtil
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.button.MaterialButton

class GameBoard : AppCompatActivity() {

    private companion object {
        const val AI_MOVE_DELAY_MS = 400L
        const val GAME_OVER_DIALOG_DELAY_MS = 700L
    }

    private lateinit var viewModel: GameBoardViewModel
    private lateinit var preferences: GamePreferences
    private lateinit var soundManager: SoundManager
    private lateinit var hapticManager: HapticManager

    /** Game state, owned by the [viewModel] so it survives configuration changes. */
    private val game: TicTacToeGame
        get() = viewModel.game

    private lateinit var cellViews: Array<ImageView>
    private lateinit var currentPlayerText: TextView
    private lateinit var xPointText: TextView
    private lateinit var oPointText: TextView
    private lateinit var drawPointText: TextView
    private lateinit var playerXTitle: TextView
    private lateinit var playerOTitle: TextView
    private lateinit var winningLineView: WinningLineView
    private lateinit var btnSoundToggle: ImageButton

    private val mainHandler = Handler(Looper.getMainLooper())
    private var pendingAiRunnable: Runnable? = null
    private var pendingDialogRunnable: Runnable? = null
    private var activeDialog: AlertDialog? = null

    private var isInputLocked = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.board)
        WindowInsetsUtil.applySystemBarInsets(findViewById(R.id.root))

        preferences = GamePreferences(this)
        soundManager = SoundManager(this).apply {
            isSoundEnabled = preferences.isSoundEnabled
        }
        hapticManager = HapticManager(this).apply {
            isHapticEnabled = preferences.isHapticEnabled
        }

        initViews()
        setupBackNavigation()

        var mode = GameMode.TWO_PLAYERS
        intent?.getStringExtra(MainActivity.EXTRA_GAME_MODE)?.let { modeName ->
            runCatching { GameMode.valueOf(modeName) }.getOrNull()?.let { mode = it }
        }

        var humanPlayer = preferences.preferredSymbol
        intent?.getStringExtra(MainActivity.EXTRA_HUMAN_PLAYER)?.let { pName ->
            runCatching { Player.valueOf(pName) }.getOrNull()?.let { humanPlayer = it }
        }

        viewModel = ViewModelProvider(
            this,
            GameBoardViewModel.factory(mode, humanPlayer)
        )[GameBoardViewModel::class.java]

        setupToolbar()
        updateLabelsForGameMode()
        updateFullUi()

        // If it starts or was restored during an AI turn, trigger AI move
        if (game.gameMode.isAiMode && game.activePlayer == viewModel.aiPlayer && !game.gameResult.isGameOver) {
            triggerAiMove()
        }
    }

    private fun initViews() {
        cellViews = arrayOf(
            findViewById(R.id.c0), findViewById(R.id.c1), findViewById(R.id.c2),
            findViewById(R.id.c3), findViewById(R.id.c4), findViewById(R.id.c5),
            findViewById(R.id.c6), findViewById(R.id.c7), findViewById(R.id.c8)
        )

        cellViews.forEachIndexed { index, cellView ->
            cellView.setOnClickListener { onCellClicked(index) }
        }

        currentPlayerText = findViewById(R.id.current_player)
        xPointText = findViewById(R.id.x_point)
        oPointText = findViewById(R.id.o_point)
        drawPointText = findViewById(R.id.draw_point)
        playerXTitle = findViewById(R.id.player_x_title)
        playerOTitle = findViewById(R.id.player_o_title)
        winningLineView = findViewById(R.id.winning_line_view)

        findViewById<MaterialButton>(R.id.btn_play_again).setOnClickListener { startNewGameRound() }
        findViewById<MaterialButton>(R.id.btn_reset_score).setOnClickListener { confirmResetScore() }
    }

    private fun setupToolbar() {
        findViewById<MaterialToolbar>(R.id.toolbar).apply {
            setNavigationIcon(R.drawable.ic_arrow_back)
            setNavigationOnClickListener { finish() }
            title = getModeTitle()
        }

        btnSoundToggle = findViewById(R.id.btn_sound_toggle)
        updateSoundIcon()
        btnSoundToggle.setOnClickListener {
            val newState = !soundManager.isSoundEnabled
            soundManager.isSoundEnabled = newState
            preferences.isSoundEnabled = newState
            updateSoundIcon()
            hapticManager.performTap(it)
            val msg = if (newState) R.string.sound_enabled else R.string.sound_disabled
            Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
        }
    }

    private fun updateSoundIcon() {
        btnSoundToggle.setImageResource(
            if (soundManager.isSoundEnabled) R.drawable.ic_sound_on else R.drawable.ic_sound_off
        )
    }

    private fun getModeTitle(): String = when (game.gameMode) {
        GameMode.TWO_PLAYERS -> getString(R.string.mode_two_player)
        GameMode.VS_AI_EASY ->
            "${getString(R.string.mode_single_player)} (${getString(R.string.difficulty_easy)})"
        GameMode.VS_AI_MEDIUM ->
            "${getString(R.string.mode_single_player)} (${getString(R.string.difficulty_medium)})"
        GameMode.VS_AI_IMPOSSIBLE ->
            "${getString(R.string.mode_single_player)} (${getString(R.string.difficulty_impossible)})"
    }

    private fun setupBackNavigation() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                finish()
            }
        })
    }

    private fun updateLabelsForGameMode() {
        if (game.gameMode.isAiMode) {
            if (viewModel.humanPlayer == Player.X) {
                playerXTitle.setText(R.string.score_you)
                playerOTitle.setText(R.string.score_ai)
            } else {
                playerXTitle.setText(R.string.score_ai)
                playerOTitle.setText(R.string.score_you)
            }
        } else {
            playerXTitle.setText(R.string.cell_x)
            playerOTitle.setText(R.string.cell_o)
        }
    }

    private fun onCellClicked(index: Int) {
        if (isInputLocked) return
        if (game.gameResult.isGameOver) return
        if (game.gameMode.isAiMode && game.activePlayer != viewModel.humanPlayer) return
        if (game.cellAt(index) != CellState.EMPTY) return

        val movingPlayer = game.activePlayer
        if (!game.makeMove(index)) return

        soundManager.playMove()
        hapticManager.performTap(cellViews[index])

        animateCellPlacement(index, movingPlayer)
        handlePostMove()
    }

    private fun handlePostMove() {
        updateScores()
        updateTurnIndicator()

        val result = game.gameResult
        if (result.isGameOver) {
            isInputLocked = true
            handleGameOver(result)
            return
        }

        // If Single Player mode and now AI's turn
        if (game.gameMode.isAiMode && game.activePlayer == viewModel.aiPlayer) {
            triggerAiMove()
        }
    }

    private fun handleGameOver(result: GameResult) {
        // Persist game result in statistics
        preferences.recordGameResult(game.gameMode, viewModel.humanPlayer, result.winner)

        if (result.winner != null) {
            highlightWinningCellsIfAny()
            val lineColor = ContextCompat.getColor(
                this,
                if (result.winner == Player.X) R.color.colorX else R.color.colorO
            )
            winningLineView.startWinAnimation(result.winningIndices, lineColor)

            if (game.gameMode.isAiMode) {
                if (result.winner == viewModel.humanPlayer) {
                    soundManager.playWin()
                    hapticManager.performWin()
                } else {
                    soundManager.playDraw()
                    hapticManager.performTap()
                }
            } else {
                soundManager.playWin()
                hapticManager.performWin()
            }
        } else {
            // Draw
            soundManager.playDraw()
            hapticManager.performTap()
        }

        scheduleGameOverDialog()
    }

    private fun triggerAiMove() {
        isInputLocked = true
        updateTurnIndicator()

        cancelPendingAi()
        val runnable = Runnable {
            if (isFinishing || isDestroyed) return@Runnable
            if (game.gameResult.isGameOver) return@Runnable

            val bestMove = TicTacToeAi.getBestMove(game, game.gameMode, viewModel.aiPlayer)
            if (bestMove >= 0) {
                game.makeMove(bestMove)
                soundManager.playMove()
                hapticManager.performTap(cellViews[bestMove])
                animateCellPlacement(bestMove, viewModel.aiPlayer)
                isInputLocked = false
                handlePostMove()
            } else {
                isInputLocked = false
            }
        }
        pendingAiRunnable = runnable
        mainHandler.postDelayed(runnable, AI_MOVE_DELAY_MS)
    }

    private fun animateCellPlacement(index: Int, player: Player) {
        val cellView = cellViews[index]
        cellView.setImageResource(if (player == Player.X) R.drawable.x else R.drawable.o)
        cellView.contentDescription = getString(
            R.string.cell_description,
            index + 1,
            if (player == Player.X) getString(R.string.cell_x) else getString(R.string.cell_o)
        )

        cellView.scaleX = 0.4f
        cellView.scaleY = 0.4f
        cellView.alpha = 0f
        cellView.animate()
            .scaleX(1f)
            .scaleY(1f)
            .alpha(1f)
            .setDuration(220)
            .start()
    }

    private fun highlightWinningCellsIfAny() {
        game.gameResult.winningIndices?.forEach { idx ->
            val cell = cellViews[idx]
            cell.setBackgroundResource(R.drawable.cell_win_background)
            cell.animate()
                .scaleX(1.12f)
                .scaleY(1.12f)
                .setDuration(300)
                .withEndAction { cell.animate().scaleX(1f).scaleY(1f).setDuration(200).start() }
                .start()
        }
    }

    private fun scheduleGameOverDialog() {
        cancelPendingDialog()
        val runnable = Runnable {
            if (isFinishing || isDestroyed) return@Runnable
            showGameOverDialog()
        }
        pendingDialogRunnable = runnable
        mainHandler.postDelayed(runnable, GAME_OVER_DIALOG_DELAY_MS)
    }

    private fun showGameOverDialog() {
        dismissActiveDialog()

        val result = game.gameResult
        val message = when (result.status) {
            GameResult.Status.X_WON ->
                if (game.gameMode.isAiMode) {
                    if (viewModel.humanPlayer == Player.X) getString(R.string.winner_you) else getString(R.string.winner_ai)
                } else {
                    getString(R.string.winner_x)
                }
            GameResult.Status.O_WON ->
                if (game.gameMode.isAiMode) {
                    if (viewModel.humanPlayer == Player.O) getString(R.string.winner_you) else getString(R.string.winner_ai)
                } else {
                    getString(R.string.winner_o)
                }
            else -> getString(R.string.game_draw)
        }

        activeDialog = AlertDialog.Builder(this)
            .setTitle(R.string.dialog_game_over_title)
            .setMessage(message)
            .setCancelable(false)
            .setPositiveButton(R.string.play_again) { _, _ -> startNewGameRound() }
            .setNegativeButton(R.string.exit_to_menu) { _, _ -> finish() }
            .create()
            .apply { show() }
    }

    private fun startNewGameRound() {
        dismissActiveDialog()
        cancelPendingAi()
        cancelPendingDialog()

        winningLineView.clear()

        // X always makes the first move in standard Tic Tac Toe
        val nextStarter = Player.X

        game.startNewRound(nextStarter)
        isInputLocked = false

        resetCellViewsUi()
        updateFullUi()

        if (game.gameMode.isAiMode && game.activePlayer == viewModel.aiPlayer) {
            triggerAiMove()
        }
    }

    private fun resetCellViewsUi() {
        cellViews.forEachIndexed { i, cell ->
            cell.animate().cancel()
            cell.scaleX = 1f
            cell.scaleY = 1f
            cell.alpha = 1f
            cell.setImageDrawable(null)
            cell.setBackgroundResource(R.drawable.cell_background)
            cell.contentDescription = getString(R.string.cell_description, i + 1, getString(R.string.cell_empty))
        }
    }

    private fun confirmResetScore() {
        AlertDialog.Builder(this)
            .setTitle(R.string.reset_scores)
            .setMessage(R.string.reset_scores)
            .setPositiveButton(R.string.play_again) { _, _ ->
                game.resetAll(Player.X)
                startNewGameRound()
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    private fun updateFullUi() {
        updateScores()
        updateBoardUi()
        updateTurnIndicator()
    }

    private fun updateBoardUi() {
        val boardState = game.board
        for (i in 0 until TicTacToeGame.BOARD_SIZE) {
            val cell = cellViews[i]
            when (boardState[i]) {
                CellState.X -> {
                    cell.setImageResource(R.drawable.x)
                    cell.contentDescription = getString(R.string.cell_description, i + 1, getString(R.string.cell_x))
                }
                CellState.O -> {
                    cell.setImageResource(R.drawable.o)
                    cell.contentDescription = getString(R.string.cell_description, i + 1, getString(R.string.cell_o))
                }
                CellState.EMPTY -> {
                    cell.setImageDrawable(null)
                    cell.contentDescription = getString(R.string.cell_description, i + 1, getString(R.string.cell_empty))
                }
            }
            cell.setBackgroundResource(R.drawable.cell_background)
        }

        if (game.gameResult.isGameOver) {
            highlightWinningCellsIfAny()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun updateScores() {
        xPointText.text = game.xScore.toString()
        oPointText.text = game.oScore.toString()
        drawPointText.text = game.drawScore.toString()
    }

    private fun updateTurnIndicator() {
        val result = game.gameResult
        if (result.isGameOver) {
            currentPlayerText.setText(
                when (result.status) {
                    GameResult.Status.X_WON ->
                        if (game.gameMode.isAiMode) {
                            if (viewModel.humanPlayer == Player.X) R.string.winner_you else R.string.winner_ai
                        } else {
                            R.string.winner_x
                        }
                    GameResult.Status.O_WON ->
                        if (game.gameMode.isAiMode) {
                            if (viewModel.humanPlayer == Player.O) R.string.winner_you else R.string.winner_ai
                        } else {
                            R.string.winner_o
                        }
                    else -> R.string.game_draw
                }
            )
            return
        }

        currentPlayerText.setText(
            if (game.gameMode.isAiMode) {
                if (game.activePlayer == viewModel.aiPlayer) R.string.turn_ai else R.string.turn_you
            } else {
                if (game.activePlayer == Player.X) R.string.turn_x else R.string.turn_o
            }
        )
    }

    private fun cancelPendingAi() {
        pendingAiRunnable?.let { mainHandler.removeCallbacks(it) }
        pendingAiRunnable = null
    }

    private fun cancelPendingDialog() {
        pendingDialogRunnable?.let { mainHandler.removeCallbacks(it) }
        pendingDialogRunnable = null
    }

    private fun dismissActiveDialog() {
        activeDialog?.takeIf { it.isShowing }?.dismiss()
        activeDialog = null
    }

    override fun onDestroy() {
        cancelPendingAi()
        cancelPendingDialog()
        dismissActiveDialog()
        winningLineView.clear()
        soundManager.release()
        super.onDestroy()
    }
}